### ISSUE TITLE
Small vocab errors?

### ISSUE THREAD

**Issue by wiseman** [karpathy/char-rnn#15]
I have 284 MB of text, but it uses only 27 characters and I'm wondering if that's the cause of some issues I'm seeing.

With the default flags, I run into this error:

```
[wiseman@Intercept char-rnn (master)]$ th train.lua -data_dir data/doj
using CUDA on GPU 0...	
loading data files...	
cutting off end of data so that the batches/sequences divide evenly	
/Users/wiseman/torch/install/bin/luajit: ./util/CharSplitLMMinibatchLoader.lua:32: bad argument #2 to 'sub' (out of range at /Users/wiseman/torch/pkg/torch/generic/Tensor.c:302)
stack traceback:
	[C]: in function 'sub'
	./util/CharSplitLMMinibatchLoader.lua:32: in function 'create'
	train.lua:74: in main chunk
	[C]: in function 'dofile'
	...eman/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
	[C]: at 0x010e3cb330
```

If I turn down the `batch_size` and `seq_length`, I don't get an error but I get no output:
```
[wiseman@Intercept char-rnn (master)]$ th train.lua -data_dir data/doj -batch_size 10 -seq_length 20
using CUDA on GPU 0...	
loading data files...	
cutting off end of data so that the batches/sequences divide evenly	
reshaping tensor...	
data load done. Number of batches in train: 0, val: 0, test: 1	
vocab size: 27	
creating an LSTM with 2 layers	
number of parameters in the model: 135127	
cloning criterion	
cloning softmax	
cloning embed	
cloning rnn	
```

If I turn the sizes down further, I get a checkpoint file with a `val_bpc` of NaN:
```
[wiseman@Intercept char-rnn (master)]$ time th train.lua -data_dir data/doj -batch_size 10 -seq_length 10
time th train.lua -data_dir data/doj -batch_size 10 -seq_length 10
using CUDA on GPU 0...	
loading data files...	
cutting off end of data so that the batches/sequences divide evenly	
reshaping tensor...	
data load done. Number of batches in train: 2, val: 0, test: 1	
vocab size: 27	
creating an LSTM with 2 layers	
number of parameters in the model: 135127	
cloning criterion	
cloning softmax	
cloning embed	
cloning rnn	
1/60 (epoch 0.500), train_bpc = 4.76364329, grad/param norm = 9.2436e-02, time/batch = 0.05s	
2/60 (epoch 1.000), train_bpc = 4.69448570, grad/param norm = 1.1961e-01, time/batch = 0.03s	
[...]
58/60 (epoch 29.000), train_bpc = 3.57610811, grad/param norm = 2.2238e-01, time/batch = 0.02s	
59/60 (epoch 29.500), train_bpc = 3.95961556, grad/param norm = 2.4135e-01, time/batch = 0.02s	
evaluating loss over split index 2	
saving checkpoint to cv/lm_lstm_epoch30.00_nan.t7	
60/60 (epoch 30.000), train_bpc = 3.68077448, grad/param norm = 3.6531e-01, time/batch = 0.02s	
```

**Comment by karpathy** [comment]
See this line:

`data load done. Number of batches in train: 2, val: 0, test: 1  `

Somehow only 3 distinct batches of data were read in, and validation set has 0 (which is why you're seeing NaN - there is no validation data). 

This makes it seem like you have a very tiny dataset, and that it's somehow not being read in correctly.

**Comment by wiseman** [comment]
Ah, thanks for the pointer.  I had inadvertently started training with a very small input.txt, and did not realize I had to delete the *.t7 files once I switched to a new, larger dataset.  Once I deleted those, training worked as expected.  (Though now I'm thinking back to other times I started with smaller datasets and then retrained with a larger set without deleting those files... ouch.)

**Comment by karpathy** [comment]
Doh, see I tried to be clever and cache the files, I didn't think people would be swapping out the data in place. Hmmmmmm. Interesting, thanks for reporting back. Maybe I should do some checksumy things.
